### PR TITLE
raftstore: fix a panic about read index during transfer leader

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

###  What have you changed?
For the panic detail, take a look at https://github.com/tikv/tikv/issues/7104.

When a leader peer pops a pending read request from `read_queue`, the peaked one may have a different ID from the ReadRequest from `raft::Ready`. It's because resolved entries in `read_queue` won't be cleared on role change, so it's possible to fetch requests proposed in the last term.

This PR fixes it by handling those old requests in the last term.

###  What is the type of the changes?
Bugfix.

###  How is the PR tested?
- Unit test

###  Any examples? (optional)

### CODE REVIEW

**Comment by BusyJay** [comment]
Can you explain more? What's the panic? How is it triggered? And how does this PR fix it?

File: components/raftstore/src/store/read_queue.rs [hunk 1]
```diff
@@ -164,12 +164,17 @@ impl ReadIndexQueue {
     }
 
     /// A fast path for `advance_leader_reads` then `pop_front`.
-    pub fn advance_leader_read_and_pop(&mut self, uuid: Uuid, index: u64) -> ReadIndexRequest {
-        let mut read = self.reads.pop_front().unwrap();
-        self.handled_cnt += 1;
-        assert_eq!(uuid, read.id);
-        read.read_index = Some(index);
-        read
+    pub fn advance_leader_read_and_pop(&mut self, uuid: Uuid, index: u64) -> Vec<ReadIndexRequest> {
```
Comment (BusyJay) [inline components/raftstore/src/store/read_queue.rs]: Does `advance_leader_reads` have the same problem?

Comment (hicqu) [reply components/raftstore/src/store/read_queue.rs]: No because `ready_cnt` is considered in the function. I add a new test case to check it.

Comment (BusyJay) [reply components/raftstore/src/store/read_queue.rs]: In most case, there should be only one entry. Is there a way to omit the allocation?

File: components/raftstore/src/store/peer.rs [hunk 2]
```diff
@@ -1486,22 +1486,24 @@ impl Peer {
...
+            if !self.is_leader() {
+                // NOTE: there could still be some pending reads proposed by the peer when it was
+                // leader. They will be cleared in `clear_uncommitted_on_role_change` later in
+                // the function.
+                self.pending_reads.advance_replica_reads(states);
+                self.post_pending_read_index_on_replica(ctx);
+            } else {
+                self.pending_reads.advance_leader_reads(states);
+                propose_time = Some(self.pending_reads.last_ready().unwrap().renew_lease_time);
+                if self.ready_to_handle_read() {
```
Comment (BusyJay) [inline components/raftstore/src/store/peer.rs]: In the past, this Line may be executed even `read_states` is empty.

**Pull request merged**